nailgun/entity_mixins.py

```
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        For example, to ask for all :class:`nailgun.entities.Organization`
+        entities::
+
+            Organization().search()
+
+        Values on an entity are used to generate a search query, and the
+        ``fields`` argument can be used to specify which fields should be used
+        when generating a search query::
+
+            org = Organization(name='foo', description='bar')
+            results = org.search()  # Search by name and description.
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
+            results = org.search(fields={'title'})  # Error!
+
+        In some cases, the simple search queries that can be generated by
+        NailGun are not sufficient. In this case, you can pass in a raw search
+        query instead. For example, to search for all organizations with a name
+        of 'foo'::
+
+            Organization().search(query={'search': u'name="foo"'})
+
+        The example above is rather pointless: it is easier and more concise to
+        use a generated query. But — and this is a **very** important "but" —
+        the manual search query is melded in to the generated query. This can
+        be used to great effect::
+
+            Organization(name='foo').search(query={'per_page': 50})
+
+        In some cases, the server's search facilities may be insufficient, or
+        it may be inordinately difficult to craft a search query. In this case,
+        you can filter search results locally. For example, to ask the server
+        for a list of all organizations and then locally search through the
+        results for the organization named "foo"::
+
+            Organization().search(filters={'name': 'foo'})
```

`filters` performs exact matching. The highlighted code above has the same effect as this:

```python
orgs = entities.Organization().search()
orgs = [org for org in orgs if org.name == 'foo']
```

This:

```python
Organization(name='foo').search()
```

...passes a search term to the server and lets the server conduct the search. The latter is vastly more efficient.

Unfortunately, `search_payload` is very naive right now. Here's what it'll do:

```python
>>> from nailgun import entities
>>> entities.Organization(name='foo').search_payload()
{'name': 'foo'}
```

And unfortunately, that search will fail. It'll fail silently, in fact - Satellite will respond to this search by returning **every single organization**.